src/core/io/platform/f4xx/ADCf4xx.cpp

```diff
@@ -84,109 +127,222 @@ void ADCf4xx::initADC(uint8_t num_channels) {
     /** Configure the global features of the ADC (Clock, Resolution, Data
      * Alignment and number of conversion)
      */
-    halADC.Instance                   = ADC1;
-    halADC.Init.ClockPrescaler        = ADC_CLOCK_SYNC_PCLK_DIV2;
-    halADC.Init.Resolution            = ADC_RESOLUTION_12B;
-    halADC.Init.ScanConvMode          = ENABLE;
-    halADC.Init.ContinuousConvMode    = ENABLE;
-    halADC.Init.DiscontinuousConvMode = DISABLE;
-    halADC.Init.ExternalTrigConvEdge  = ADC_EXTERNALTRIGCONVEDGE_NONE;
-    halADC.Init.ExternalTrigConv      = ADC_SOFTWARE_START;
-    halADC.Init.DataAlign             = ADC_DATAALIGN_RIGHT;
-    halADC.Init.NbrOfConversion       = num_channels;
-    halADC.Init.DMAContinuousRequests = ENABLE;
-    halADC.Init.EOCSelection          = ADC_EOC_SEQ_CONV;
-
-    __HAL_RCC_ADC1_CLK_ENABLE();
-    HAL_ADC_Init(&halADC);
+    ADCf4xx::ADC_State_t* adcState = &adcArray[getADCNum()];
+    ADC_HandleTypeDef* halADC      = &adcState->halADC;
+    // Set instance to the ADC peripheral being using
+    halADC->Init.ClockPrescaler        = ADC_CLOCK_SYNC_PCLK_DIV2;
+    halADC->Init.Resolution            = ADC_RESOLUTION_12B;
+    halADC->Init.ScanConvMode          = ENABLE;
+    halADC->Init.ContinuousConvMode    = DISABLE;
+    halADC->Init.DiscontinuousConvMode = DISABLE;
+    halADC->Init.ExternalTrigConvEdge  = ADC_EXTERNALTRIGCONVEDGE_RISING;
+    halADC->Init.ExternalTrigConv      = ADC_EXTERNALTRIGCONV_T8_TRGO;
+    halADC->Init.DataAlign             = ADC_DATAALIGN_RIGHT;
+    halADC->Init.NbrOfConversion       = num_channels;
+    halADC->Init.DMAContinuousRequests = ENABLE;
+    halADC->Init.EOCSelection          = ADC_EOC_SINGLE_CONV;
+
+    switch (getADCNum()) {
+    case ADC1_SLOT:
+        halADC->Instance = ADC1;
+        __HAL_RCC_ADC1_CLK_ENABLE();
+        break;
+    case ADC2_SLOT:
+        halADC->Instance = ADC2;
+        __HAL_RCC_ADC2_CLK_ENABLE();
+        break;
+    case ADC3_SLOT:
+        halADC->Instance = ADC3;
+        __HAL_RCC_ADC3_CLK_ENABLE();
+        break;
+    default:
+        return; // Should never get here
+    }
+    HAL_ADC_Init(halADC);
```

Comment:
every time, because every time we add a channel / instance, we need to initialize the ADC peripheral with a new  number of conversions (1 higher). The number of conversions is the the num of channels on that specific adc.